### PR TITLE
feat: t11 - chapter3-t11-php 

### PR DESCRIPTION
**Pull request by darliiin**
# Summary
## Test Task
**slow php reporting page**
A typical scenario:
- you have a web app
- that app has a backend written in PHP
- most of the backend works fast and good for end users < 1000ms
- however there is a slow page /reports/slow, that takes 10m to execute, which causes nginx and php to timeout
- implement such solution so that nginx/php-fpm do not timeout and get the report when 10m passed.

Implementation nuances:
- your nginx and php-fpm should run under nginx/nginx user/group.
- use sleep php func to simulate such report, https://www.php.net/manual/en/function.sleep.php
- implement a custom location /reports/fast, that will work fast and doesn't require any timeouts, and will use www pool of the PHP-FPM
- implement a custom location /reports/slow, that will open /tmp/nginx/reports/slow.php script, which will take 10m to execute
- during the /reports/slow, you should not lose the browser connection to the server.
- IMPORTANT: you can't change any values for your core application that runs inside www pool defined in the PHP-FPM, for example you can't change timeout via www pool, as that may substantially degrade all the other URLs that are working good and fast - so think how you can do it.

When you open PR - pls package your solution as zip file with bash script run.sh that should
- create all necessary files and/or folders.
- set proper permissions to created files/folders
- start nginx in foreground mode (so all errors, logs are printed on stdout/stderr)
- start php-fpm in foreground mode (so all errors, logs are printed on stdout/stderr)
- curl https://localhost/reports/fast should output "FAST REPORT" via PHP script
- curl https://localhost/reports/slow should output "SLOW REPORT" via PHP script with 10m delay without losing any connection (i.e. no timeout or upstream timeout error in the nginx)

# My results
```
curl https://localhost/reports/fast
curl https://localhost/reports/slow
```
**output:**
Screenshot_20230427_000714
[t11.zip](https://github.com/darliiin/devops-camp-2023/files/11298642/t11.zip)

### CODE REVIEW

File: chapter3/tests/t11/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	access_log /var/log/nginx/access.log json;
+
+	sendfile on;
+	tcp_nopush on;
+	types_hash_max_size 2048;
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+
+	upstream reports {
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/nginx.conf:31]: that is not enough you need to have a dedicated php-fpm pool in which you set timeout for php scripts via `php_admin` call.

and your slow page should use that php-fpm 'slow' pool

File: chapter3/tests/t11/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			fastcgi_param  SERVER_SOFTWARE	nginx/$nginx_version;
+			fastcgi_param  REMOTE_ADDR	$remote_addr;
+			fastcgi_param  REMOTE_PORT	$remote_port;
+			fastcgi_param  SERVER_ADDR	$server_addr;
+			fastcgi_param  SERVER_PORT	$server_port;
+			fastcgi_param  SERVER_NAME	$server_name;
+			fastcgi_param  REDIRECT_STATUS	200;
+			fastcgi_param  PATH_INFO	$fastcgi_path_info;
+
+			fastcgi_read_timeout 600s;
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/nginx.conf:108]: that is not enough you need to have a dedicated php-fpm pool in which you set timeout for php scripts via `php_admin` call.

for example your worker will wait for 10m - great, but your php.ini has value that scripts should timeout after 30s. 

That's why you need a dedicated pool used by a specific location `/reports/slow`

File: chapter3/tests/t11/php-fpm.conf [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[global]
+pid = /run/php/php8.1-fpm.pid
+
+error_log = /var/log/php8.2-fpm.error.log
+log_level = error
+
+daemonize = no
+events.mechanism = epoll
+
+[www]
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/php-fpm.conf:10]: need dedicated `[slow]` pool

File: chapter3/tests/t11/php-fpm.conf [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+user = nginx
+group = nginx
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/php-fpm.conf:12]: this should be in the global

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
prove that your connection stayed 10m with `time curl URL`

File: chapter3/tests/t11/php-fpm.conf (hunk 4 above)
Comment (darliiin) [reply chapter3/tests/t11/php-fpm.conf:11]: done

File: chapter3/tests/t11/nginx.conf (hunk 2 above)
Comment (darliiin) [reply chapter3/tests/t11/nginx.conf:112]: done

File: chapter3/tests/t11/nginx.conf (hunk 1 above)
Comment (darliiin) [reply chapter3/tests/t11/nginx.conf:31]: done

File: chapter3/tests/t11/php-fpm.conf [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+pm.max_children = 1
+
+[slow]
+listen = /run/php/slow.sock
+listen.owner = nginx
+listen.group = nginx
+
+pm = static
+pm.max_children = 1
+php_admin_value[execution_time_limit] = 600
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/php-fpm.conf:28]: how  did you test? it shouuld complain on this value - there is no such setting in php.ini

File: chapter3/tests/t11/run.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+sudo chown -R nginx:nginx /tmp/localhost+1-key.pem
+
+sudo kill $(sudo lsof -t -i :443)
+
+cd -
+sudo /usr/sbin/php-fpm8.1 --fpm-config "$(pwd)/php-fpm.conf" --php-ini "$(pwd)/php.ini" > /dev/null &
+sudo nginx -c "$(pwd)/nginx.conf"
+
+curl https://localhost/reports/fast
+curl https://localhost/reports/slow
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/run.sh:15]: use `time`, i.e. `time curl URL`

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/tests/t11/run.sh (hunk 6 above)
Comment (darliiin) [reply chapter3/tests/t11/run.sh]: done

File: chapter3/tests/t11/php-fpm.conf (hunk 5 above)
Comment (darliiin) [reply chapter3/tests/t11/php-fpm.conf:28]: updated php ini

Comment (dmitry-mightydevops) [reply chapter3/tests/t11/php-fpm.conf:28]: you should not update php.ini - you should use proper value for php_admin_value in the php-fpm.conf

File: chapter3/tests/t11/php.ini [hunk 7]
```diff
@@ -0,0 +1,1143 @@
...
+; Miscellaneous ;
+;;;;;;;;;;;;;;;;;
+
+expose_php = Off
+
+;;;;;;;;;;;;;;;;;;;
+; Resource Limits ;
+;;;;;;;;;;;;;;;;;;;
+
+php_admin_value max_execution_timeout = 30s
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/php.ini:35]: incorrect.

Comment (dmitry-mightydevops) [reply chapter3/tests/t11/php.ini:35]: you have slow and fast php-fpm pools
fast - should limit execution time to 60s
slow - to 600s

that should not be defined in php.ini

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/tests/t11/php.ini (hunk 7 above)
Comment (darliiin) [reply chapter3/tests/t11/php.ini]: corrected

File: chapter3/tests/t11/php-fpm.conf [hunk 8]
```diff
@@ -25,4 +25,4 @@ listen.group = nginx
 
 pm = static
 pm.max_children = 1
-php_admin_value[execution_time_limit] = 600
+php_admin_value[max_execution_time] = 600
```
Comment (dmitry-mightydevops) [inline chapter3/tests/t11/php-fpm.conf:28]: you should have one in the [www] pool to limit to 60s

Comment (dmitry-mightydevops) [reply chapter3/tests/t11/php-fpm.conf:28]: please perform the change and show me `time ` output for both URls

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/tests/t11/php-fpm.conf (hunk 8 above)
Comment (darliiin) [reply chapter3/tests/t11/php-fpm.conf:30]: Screenshot_20230430_132844

**Review by dmitry-mightydevops — APPROVED** [review]

**Pull request merged**